src/quacc/recipes/torchsim/core.py

```diff
@@ -0,0 +1,364 @@
+    convergence_fn_kwargs: dict[str, Any] | None = None,
+    **optimizer_kwargs: Any,
+) -> TorchSimOptSchema:
+    """
+    Carry out a geometry optimization on a set of atoms.
+
+    Parameters
+    ----------
+    atoms : list[Atoms]
+        The list of atoms objects.
+    model_type : TSModelType
+        The type of model to use, limited to the types supported by TorchSim.
+    model_path : str | Path
+        The path to the model file or checkpoint.
+    optimizer : Optimizer
+        The TorchSim optimizer to use.
+    convergence_fn : ConvergenceFn
+        The convergence function, either "energy" or "force". This will use either the
+        ts.generate_energy_convergence_fn or ts.generate_force_convergence_fn function
+        to interally generate the convergence function. Arguments can be supplied via
+        the convergence_fn_kwargs argument.
+    trajectory_reporter_dict : TrajectoryReporterDict | None
+        This dictionary defines the trajectory reporting behavior. This is a
+        quacc-specific dictionary that allows for the configuration of the
+        TrajectoryReporter. For a list of available keys, refer to the TorchSim
+        TrajectoryReporter documentation or the TrajectoryReporterDict type-hint.
+    autobatcher_dict : AutobatcherDict | bool
+        This dictionary defines the autobatcher behavior. This is a quacc-specific
+        dictionary that allows for the configuration of the autobatcher. For a list of
+        available keys, refer to the TorchSim Autobatcher's documentation or the
+        AutobatcherDict type-hint.
+    max_steps : int
+        The maximum number of steps to run for each optimization.
+    steps_between_swaps : int
+        Number of steps to take before checking convergence and swapping out states.
+    init_kwargs : dict[str, Any] | None
+        Keyword arguments passed to the optimizer initialization function.
+    model_kwargs : dict[str, Any] | None
+        Keyword arguments passed to the model.
+    convergence_fn_kwargs : dict[str, Any] | None
+        Keyword arguments passed to the convergence function generator.
+    **optimizer_kwargs : Any
+        Keyword arguments to pass to the optimizer step function.
+
+    Returns
+    -------
+    TorchSimOptSchema
+        A dictionary representing the final atoms configuration and metadata geometry
+        optimization schema. See the type-hint for the data structure.
+    """
+    runner = BaseRunner()
+    runner.setup()
+
+    model = pick_model(model_type, model_path, **model_kwargs or {})
+
+    convergence_fn_obj = CONVERGENCE_FN_REGISTRY[convergence_fn](
+        **convergence_fn_kwargs or {}
+    )
+
+    state = ts.initialize_state(atoms, model.device, model.dtype)
+
+    # Configure trajectory reporter
+    trajectory_reporter_dict = trajectory_reporter_dict or {}
+    trajectory_reporter_dict["filenames"] = [
+        runner.tmpdir / f"trajectory_{i}.h5md" for i in range(len(atoms))
+    ]
```

Comment:
Here, we ensure that all files generated at runtime are being written out to the unique temporary directory. It is important that quacc controls all file I/O because that is the only way to prevent concurrency issues. For instance, if the user decides to run 100 jobs in the current working directory without any sort of workflow engine, we want to ensure that quacc prevents those jobs from clashing with one another. That is why quacc handles the file handling. In the case of a workflow engine that handles directory management (e.g. FireWorks, Jobflow-Remote), quacc will let the workflow engine handle it all instead. But not all workflow engines support this, and we need to ensure that someone using no workflow engine will still be able to launch jobs without concern.